clippy_lints/src/methods/unnecessary_fallible_conversions.rs

```diff
@@ -45,47 +46,80 @@ fn check<'tcx>(
+        // If there is an unwrap/expect call, extend the span to include the call
+        let span = if let Some(unwrap_call) = parent_unwrap_call {
+            primary_span.with_hi(unwrap_call.hi())
+        } else {
+            primary_span
+        };
+
+        let qpath_spans = qpath.and_then(|qpath| match qpath {
+            QPath::Resolved(_, path) => {
+                let segments = path.segments.iter().map(|seg| seg.ident).collect::<Vec<_>>();
+                (segments.len() == 2).then(|| vec![segments[0].span, segments[1].span])
+            },
+            QPath::TypeRelative(_, seg) => Some(vec![seg.ident.span]),
+            QPath::LangItem(_, _) => unreachable!("`TryFrom` and `TryInto` are not lang items"),
+        });
+
+        let (source_ty, target_ty, sugg, applicability) = match (kind, &qpath_spans, parent_unwrap_call) {
+            (FunctionKind::TryIntoMethod, _, Some(unwrap_span)) => {
+                let sugg = vec![(primary_span, String::from("into")), (unwrap_span, String::new())];
+                (self_ty, other_ty, sugg, Applicability::MachineApplicable)
+            },
+            (FunctionKind::TryFromFunction, Some(spans), Some(unwrap_span)) => {
+                let sugg = match spans.len() {
+                    1 => vec![(spans[0], String::from("from")), (unwrap_span, String::new())],
+                    2 => vec![
+                        (spans[0], String::from("From")),
+                        (spans[1], String::from("from")),
+                        (unwrap_span, String::new()),
+                    ],
+                    _ => unreachable!(),
+                };
+                (other_ty, self_ty, sugg, Applicability::MachineApplicable)
+            },
+            (FunctionKind::TryIntoFunction, Some(spans), Some(unwrap_span)) => {
+                let sugg = match spans.len() {
+                    1 => vec![(spans[0], String::from("into")), (unwrap_span, String::new())],
+                    2 => vec![
+                        (spans[0], String::from("Into")),
+                        (spans[1], String::from("into")),
+                        (unwrap_span, String::new()),
+                    ],
+                    _ => unreachable!(),
+                };
+                (self_ty, other_ty, sugg, Applicability::MachineApplicable)
+            },
+            (FunctionKind::TryFromFunction, _, _) => {
+                let sugg = vec![(primary_span, String::from("From::from"))];
+                (other_ty, self_ty, sugg, Applicability::Unspecified)
+            },
+            (FunctionKind::TryIntoFunction, _, _) => {
+                let sugg = vec![(primary_span, String::from("Into::into"))];
+                (self_ty, other_ty, sugg, Applicability::Unspecified)
+            },
+            (FunctionKind::TryIntoMethod, _, _) => {
+                let sugg = vec![(primary_span, String::from("into"))];
+                (self_ty, other_ty, sugg, Applicability::Unspecified)
             },
```

Comment:
Could we simplify this by adding a method to `FunctionKind` for the suggestion? Something like:
`(other_ty, self_ty, kind.sugg(), Applicability::Unspecified)`.

This applies to `self_ty` and `other_ty` as well, not bc of code duplication but just that it's hard to understand what they are (without scrolling up and seeing `source_ty` and `target_ty`). And the entire match should be extracted tbh

Overall this entire match is a bit of a mess, once I have my desktop back (probably this weekend) I'll take a closer look and try my hand at it. It's pretty hard to understand in its current form.